core/unix/signal.c

```diff
@@ -3064,6 +3068,26 @@ fixup_rtframe_pointers(dcontext_t *dcontext, int sig, sigframe_rt_t *f_old,
                    sizeof(xstate_new->xstate_hdr));
             memcpy(&xstate_new->ymmh, &xstate_old->ymmh, sizeof(xstate_new->ymmh));
         }
+#    ifdef X64
+        if (ZMM_ENABLED()) {
+            kernel_xstate_t *xstate_new = (kernel_xstate_t *)tgt;
+            kernel_xstate_t *xstate_old =
+                (kernel_xstate_t *)f_old->uc.uc_mcontext.fpstate;
+            memcpy((byte *)xstate_new + proc_xstate_area_zmm_hi256_offs(),
+                   (byte *)xstate_old + proc_xstate_area_zmm_hi256_offs(),
+                   proc_num_simd_sse_avx_registers() * ZMMH_REG_SIZE);
+            memcpy((byte *)xstate_new + proc_xstate_area_hi16_zmm_offs(),
+                   (byte *)xstate_old + proc_xstate_area_hi16_zmm_offs(),
+                   proc_num_simd_sse_avx_registers() * ZMM_REG_SIZE);
+            memcpy((byte *)xstate_new + proc_xstate_area_kmask_offs(),
+                   (byte *)xstate_old + proc_xstate_area_kmask_offs(),
+                   proc_num_opmask_registers() * OPMASK_AVX512BW_REG_SIZE);
+        }
+#    else
+        /* XXX i#1312: it is unclear if and how the components are arranged in
```

Comment:
So is this a fatal issue?  This seems stronger than just an XXX comment?

Reply:
I wouldn't make it fatal. I'm actually not sure (I think it may not) if the kernel provides the context in 32-bit. Experimentally it looked like it may not. As I said above, I do see your point of making it stronger than just ignoring with a comment. Again same question, WDYT about SYSLOG_INTERNAL_ERROR_ONCE?

Reply:
I don't see how the kernel can not save the zmmh for 32-bit.  If the processor supports zmmh access in 32-bit, would the kernel really not preserve it?

Reply:
Explaining my wording there: this is not about just presenting a picture of state to a signal handler.  This is *the* copy of the state used to resume execution when the signal handler is done.  If zmmh is not here, it's not preserved by the kernel.

Reply:
Where would you put this SYSLOG?  Under 32-bit mode with ZMM_ENABLED?  So it will fire every single time for a 32-bit process on a new processor?

Reply:
We always have the state in our mcontext, even for non-pending signals right? We never just copy the signal frame as a whole to the stack w/o examining it..? So yes, it's basically unsupported in 32-bit and no copy is made yet until I know if and what the kernel does (and the xstate instruction by the way) in 32-bit. That's why I did not add any error or warning message and instead explained it in the comments. It's simply unsupported, no change from status quo for 32-bit.

Reply:
Not sure what you mean about having state in the mcontext.  This is a fundamental issue: if the kernel does not place zmmh state in signal frames, then it doesn't preserve zmmh state at all, and thus 32-bit apps cannot use zmmh at all on Linux.  DR should probably then remove zmmh from mcontext_t, context switch code, clean calls, etc. for 32-bit Linux.

However, that does not sound likely to me.  This seems like a missing piece of AVX-512 support in DR.  It rates more than an `XXX` comment which is a nice-to-have that could be skipped: this is a correctness issue, a `FIXME`.  We're going to corrupt zmmh state on sigreturn in 32-bit with uninitialized data.

Reply:
Yes, definitely FIXME, I actually had misunderstood the meaning of XXX. This is a comment bug.

Reply:
I've changed all the XXX comments about this to FIXME. The commit message already mentions it.

Reply:
> I don't see how the kernel can not save the zmmh for 32-bit. If the processor supports zmmh access in 32-bit, would the kernel really not preserve it?

Likely it is there but it is absolutely not clear where. The straight forward way as described in the Intel manuals to query cpuid about the offsets does not apply on the machines I've tried. It's possible AVX-512 was somehow disabled in 32-bit compatibility mode on my machines? But XCR0 seemed like it was enabled. Need to find a kernel person about this.